src/minifluxapi.cpp

```diff
@@ -0,0 +1,318 @@
+	// as read
+	rsspp::Feed feed = fetch_feed(id, nullptr);
+
+	std::vector<std::string> guids;
+	for (const auto& item : feed.items) {
+		guids.push_back(item.guid);
+	}
+
+	json args;
+	args["status"] = "read";
+
+	return update_articles(guids, args);
+}
+
+bool MinifluxApi::mark_article_read(const std::string& guid, bool read)
+{
+	// Do this in a thread, as we don't care about the result enough to wait
+	// for it.
+	std::thread t{[=]()
+	{
+		LOG(Level::DEBUG,
+			"MinifluxApi::mark_article_read: inside thread, marking "
+			"article as read...");
+
+		// Call the MinifluxApi's update_article function as a thread.
+		json args;
+		args["status"] = read ? "read" : "unread";
+		this->update_article(guid, args);
+	}};
+	t.detach();
+	return true;
+}
+
+bool MinifluxApi::flag_changed(const std::string& oldflags,
+	const std::string& newflags,
+	const std::string& flagstr)
+{
+	if (flagstr.length() == 0) {
+		return false;
+	}
+
+	const char flag = flagstr[0];
+	const char* oldptr = strchr(oldflags.c_str(), flag);
+	const char* newptr = strchr(newflags.c_str(), flag);
+
+	if (oldptr == nullptr && newptr != nullptr) {
+		return true;
+	}
+	if (oldptr != nullptr && newptr == nullptr) {
+		return true;
+	}
+
+	return false;
+}
+
+bool MinifluxApi::update_article_flags(const std::string& oldflags,
+	const std::string& newflags,
+	const std::string& guid)
+{
+	std::string star_flag = cfg->get_configvalue("miniflux-flag-star");
+	bool starred_flag_changed = flag_changed(oldflags, newflags, star_flag);
+
+	bool success = true;
+	if (starred_flag_changed) {
+		success = toggle_star_article(guid);
+	}
+
+	return success;
+}
+
+rsspp::Feed MinifluxApi::fetch_feed(const std::string& id, CURL* cached_handle)
+{
+	rsspp::Feed feed;
+	feed.rss_version = rsspp::Feed::MINIFLUX_JSON;
+
+	std::string query =
+		strprintf::fmt("/v1/feeds/%s/entries?order=published_at&direction=desc", id);
+
+	json content = run_op(query, json(), "GET", cached_handle);
+	if (content.is_null()) {
+		return feed;
+	}
+
+	json entries = content["entries"];
+	if (!entries.is_array()) {
+		LOG(Level::ERROR,
+			"MinifluxApi::fetch_feed: items is not an array");
+		return feed;
+	}
+
+	LOG(Level::DEBUG,
+		"MinifluxApi::fetch_feed: %" PRIu64 " items",
+		static_cast<uint64_t>(entries.size()));
+	try {
+		for (const auto& entry : entries) {
+			rsspp::Item item;
+
+			if (!entry["title"].is_null()) {
+				item.title = entry["title"];
+			}
+
+			if (!entry["url"].is_null()) {
+				item.link = entry["url"];
+			}
+
+			if (!entry["author"].is_null()) {
+				item.author = entry["author"];
+			}
+
+			if (!entry["content"].is_null()) {
+				item.content_encoded = entry["content"];
+			}
+
+			int entry_id = entry["id"];
+			item.guid = std::to_string(entry_id);
+
+			item.pubDate = entry["published_at"];
+
+			std::string status = entry["status"];
+			if (status == "unread") {
+				item.labels.push_back("miniflux:unread");
+			} else {
+				item.labels.push_back("miniflux:read");
+			}
```

Comment:
Hmm good point. It looks like there's no way to set the flags from here, though, because these Item objects don't store any flag information (since flags are specific to Newsboat and not RSS in general). Unless I'm missing something most of the other backends also don't handle this properly (OCNews creates a virtual tag with all the starred articles). I'm wondering now if this needs a deeper rework, so that backends can correctly set the flag on articles they download; maybe adding a "starred" label that gets parsed further up?

Reply:
> I'm wondering now if this needs a deeper rework, so that backends can correctly set the flag on articles they download

That's be ideal, but most of the backends don't have a designated maintainer, so I don't have high hopes for refactors like that. The next best thing is to improve the new and open-source backends, which is what I'm trying to do here :)

> maybe adding a "starred" label that gets parsed further up?

I think this is a fine solution. We already do that for read/unread status, might as well extend that to starred items. (That's a bit kludgey, but will do for now.)

Reply:
From IRC discussion with @gjabell: setting a flag from `MinifluxApi` overwrites the flags that are already in the database, because `MinifluxApi` doesn't know which flags are already set on the item. This is a more general problem and is out of scope for this PR; I'm calling off this review comment, and will file a separate issue about this.

Reply:
Filed https://github.com/newsboat/newsboat/issues/1130.